pkg/engine/lifecycletest/pulumi_test.go

```diff
@@ -1740,6 +1740,93 @@ func TestAliases(t *testing.T) {
 		},
 	}}, []display.StepOp{deploy.OpCreate}, false)
 
+	// ensure that aliases of parents of parents resolves correctly
+	// first create a chain of resources such that we have n1 -> n1-sub -> n1-sub-sub
+	snap = updateProgramWithResource(nil, []Resource{{
+		t:    "pkgA:index:t1",
+		name: "n1",
+	}, {
+		t:      "pkgA:index:t2",
+		name:   "n1-sub",
+		parent: resource.URN("urn:pulumi:test::test::pkgA:index:t1::n1"),
+	}, {
+		t:      "pkgA:index:t3",
+		name:   "n1-sub-sub",
+		parent: resource.URN("urn:pulumi:test::test::pkgA:index:t1$pkgA:index:t2::n1-sub"),
+	}}, []display.StepOp{deploy.OpCreate}, false)
+
+	// Now change n1's name and type
+	_ = updateProgramWithResource(snap, []Resource{{
+		t:    "pkgA:index:t1-new",
+		name: "n1-new",
+		aliases: []resource.URN{
+			"urn:pulumi:test::test::pkgA:index:t1::n1",
+		},
+	}, {
+		t:      "pkgA:index:t2",
+		name:   "n1-new-sub",
+		parent: resource.URN("urn:pulumi:test::test::pkgA:index:t1-new::n1-new"),
+		aliases: []resource.URN{
+			"urn:pulumi:test::test::pkgA:index:t1$pkgA:index:t2::n1-sub",
+		},
+	}, {
+		t:      "pkgA:index:t3",
+		name:   "n1-new-sub-sub",
+		parent: resource.URN("urn:pulumi:test::test::pkgA:index:t1-new$pkgA:index:t2::n1-new-sub"),
+		aliases: []resource.URN{
+			"urn:pulumi:test::test::pkgA:index:t1$pkgA:index:t2$pkgA:index:t3::n1-sub-sub",
+		},
+	}}, []display.StepOp{deploy.OpSame}, false)
+
+	// Test catastrophic multiplication out of aliases doesn't crash out of memory
+	// first create a chain of resources such that we have n1 -> n1-sub -> n1-sub-sub
+	snap = updateProgramWithResource(nil, []Resource{{
+		t:    "pkgA:index:t1-v0",
+		name: "n1",
+	}, {
+		t:      "pkgA:index:t2-v0",
+		name:   "n1-sub",
+		parent: resource.URN("urn:pulumi:test::test::pkgA:index:t1-v0::n1"),
+	}, {
+		t:      "pkgA:index:t3",
+		name:   "n1-sub-sub",
+		parent: resource.URN("urn:pulumi:test::test::pkgA:index:t1-v0$pkgA:index:t2-v0::n1-sub"),
+	}}, []display.StepOp{deploy.OpCreate}, false)
+
+	// Now change n1's name and type and n2's type, but also add a load of aliases and pre-multiply them out
+	// before sending to the engine
+	n1Aliases := make([]resource.URN, 0)
+	n2Aliases := make([]resource.URN, 0)
+	n3Aliases := make([]resource.URN, 0)
+	for i := 0; i < 100; i++ {
+		n1Aliases = append(n1Aliases, resource.URN(
+			fmt.Sprintf("urn:pulumi:test::test::pkgA:index:t1-v%d::n1", i)))
+
+		for j := 0; j < 10; j++ {
+			n2Aliases = append(n2Aliases, resource.URN(
+				fmt.Sprintf("urn:pulumi:test::test::pkgA:index:t1-v%d$pkgA:index:t2-v%d::n1-sub", i, j)))
+
+			n3Aliases = append(n3Aliases, resource.URN(
+				fmt.Sprintf("urn:pulumi:test::test::pkgA:index:t1-v%d$pkgA:index:t2-v%d$pkgA:index:t3::n1-sub-sub", i, j)))
+		}
+	}
+
+	snap = updateProgramWithResource(snap, []Resource{{
```

Comment:
So "nXAliases" here is "really big" but the engine somehow dedups them or uses a more efficient algo? The tests regresses something that got fixed? Would it still pass on a machine with a ton of memory?

Reply:
Yeh so the engine was (before we reverted the changes) using more than expected amounts of memory trying to duplicate all these aliases as it did the same multiply out that we're doing here in the test. So yes this could pass on a machine with more memory, not sure how best to do resource usage tests in go to always catch that.